openwebstart/src/test/java/com/openwebstart/os/mac/AppFactoryTest.java

```diff
@@ -0,0 +1,180 @@
+
+			AppFactory.createApp(appName, script, IcnsFactorySample.class.getResource("icon.png").getFile());
+
+			assert Files.exists(Paths.get(userHome));
+			assert Files.exists(Paths.get(userCache));
+			assert Files.exists(Paths.get(userCache, "applications"));
+			assert Files.exists(Paths.get(userCache, "applications", appNameWithSuffix));
+
+			assert Files.exists(Paths.get(userHome, "Applications"));
+			final Path link = Paths.get(userHome, "Applications", appNameWithSuffix);
+			assert Files.exists(link);
+			assert Files.isSymbolicLink(link);
+			assert link.toRealPath().equals(Paths.get(userCache, "applications", appNameWithSuffix).toRealPath());
+
+			// do it again
+			AppFactory.createApp(appName, script, IcnsFactorySample.class.getResource("icon.png").getFile());
+		} finally {
+			restoreOrigins();
+		}
+	}
+
+	@Test
+	public void testDesktopLinkExists() throws Exception {
+		if (!OperationSystem.getLocalSystem().equals(OperationSystem.MAC64)) {
+			return;
+		}
+
+		FileUtils.recursiveDelete(new File(userHome), new File("/tmp"));
+		assert new File(userHome).exists() == false;
+
+		final String appName = "MyFirstApp";
+		final String appNameWithSuffix = appName + AppFactory.APP_EXTENSION;
+
+		try {
+			System.setProperty(JavaSystemPropertiesConstants.USER_HOME, userHome);
+			assert userHome.equals(JavaSystemProperties.getUserHome());
+			updateCacheHome(userCache);
+			assert userCache.equals(FilesystemConfiguration.getCacheHome());
+
+			assert AppFactory.desktopLinkExists(appName) == false;
+
+			final Path desktop = Paths.get(JavaSystemProperties.getUserHome(), "Desktop");
+			Files.createDirectories(desktop);
+			assert Files.isDirectory(desktop);
+
+			final Path link = desktop.resolve(appNameWithSuffix);
+
+			{ // regular file
+				assert Files.exists(link) == false;
+				Files.createFile(link);
+				assert Files.exists(link) == true;
+				assert AppFactory.desktopLinkExists(appName) == false;
+				Files.delete(link);
+			}
+
+			{ // directory
+				assert Files.exists(link) == false;
+				Files.createDirectories(link);
+				assert Files.isDirectory(link) == true;
+				assert AppFactory.desktopLinkExists(appName) == false;
+				Files.delete(link);
+			}
+
+			final Path appRoot = AppFactory.createAppWithoutMenuEntry(appName, appNameWithSuffix,
+					IcnsFactorySample.class.getResource("icon.png").getFile()).toPath();
+
+			{ // wrong link
+				assert Files.exists(link) == false;
+				Files.createSymbolicLink(link, appRoot.resolve(AppFactory.CONTENTS_FOLDER_NAME));
+				assert Files.isSymbolicLink(link) == true;
+				assert AppFactory.desktopLinkExists(appName) == false;
+				Files.delete(link);
+			}
+
+			{ // wrong link
+				assert Files.exists(link) == false;
+				Files.createSymbolicLink(link, appRoot);
+				assert Files.isSymbolicLink(link) == true;
+				assert AppFactory.desktopLinkExists(appName) == true;
+				Files.delete(link);
+			}
+		} finally {
+			restoreOrigins();
```

Comment:
We should add restoreOrigins(); in a method that is called before each test. See https://junit.org/junit5/docs/5.0.2/api/org/junit/jupiter/api/BeforeEach.html

Reply:
For each test i change user.home system property & update FilesystemConfiguration.cacheHome via reflection. This could be moved into @BeforeEach. This change may have an effect of other tests outside of AppFactoryTests. I would suggest to annotate restoreOrigins with @AfterAll.